### DISCUSSION TITLE
Websocket init function get client ip

### DISCUSSION THREAD

**Discussion by kokutas** [99designs/gqlgen#1889] [Q&A]
### What happened?
How to get ip in chat's websocket init func, set wesocket current limit, set current limit IP whitelist?
### What did you expect?
From websocket init function get client ip。
### versions
 - `gqlgen version`? 
 - github.com/99designs/gqlgen v0.14.0
 - `go version`?
 - go version go1.16.10 darwin/amd64
 - dep or go modules?
 - github.com/gorilla/websocket v1.4.2

**Comment by chedom** [comment]
Hi @kokutas ,
IP address belongs to the lower (HTTP) layer. 
You can add HTTP middleware that extracts IP address from http.Request and keeps it in Context.
```
package ip

import (
	...
	"github.com/tomasen/realip"
)

type ipCtxKey struct{}

func NewIPMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		ip := realip.FromRequest(r)
		ctx := context.WithValue(r.Context(), ipCtxKey{}, ip)
		next.ServeHTTP(w, r.WithContext(ctx))
	})
}


func FromContext(ctx context.Context) string {
	val := ctx.Value(ipCtxKey{})
	if val == nil {
		return ""
	}

	return val.(string)
}

```
WS init function  [has](https://github.com/99designs/gqlgen/blob/master/graphql/handler/transport/websocket.go#L113) access to the Context of the request.
```
func myInit(ctx context.Context, initPayload InitPayload) (context.Context, error) {
  userIP := ip.FromContext(ctx)
  ...
}
```
